source/DataWriter.h

```diff
@@ -33,86 +35,125 @@ class DataNode;
+	DataWriter &WriteComment(const std::string &str);
 
 	// Write a token, without writing a whole line. Use this very carefully.
-	void WriteToken(const char *a);
-	void WriteToken(const std::string &a);
+	DataWriter &WriteToken(const char *a);
+	DataWriter &WriteToken(const std::string &a);
 	// Write a token of any arithmetic type.
 	template <class A>
-	void WriteToken(const A &a);
+	DataWriter &WriteToken(const A &a);
+	// Writes a series of tokens without terminating the line.
+	template <class A, class ...B>
+	DataWriter &WriteToken(const A &a, B... others);
+	// Write the tokens of this DataNode without writing its children.
+	DataWriter &WriteTokens(const DataNode &node);
+
+	// Changes the separator used between tokens. The default is a single space.
+	DataWriter &SetSeparator(const std::string &sep);
+	// Changes the indentation used before child tokens. The default is a single tabulator.
+	DataWriter &SetIndentation(const std::string &indent);
 
 
 private:
-	// Save path (in UTF-8). Empty string for in-memory DataWriter.
-	std::string path;
 	// Current indentation level.
 	std::string indent;
+	// The string used for indentation.
+	std::string indentString = "\t";
 	// Before writing each token, we will write either the indentation string
 	// above, or this string.
-	static const std::string space;
+	std::string separator = " ";
 	// Remember which string should be written before the next token. This is
-	// "indent" for the first token in a line and "space" for subsequent tokens.
+	// "indent" for the first token in a line and "separator" for subsequent tokens.
+	// Null pointer signals that this string was already written for the next token.
 	const std::string *before;
 	// Compose the output in memory before writing it to file.
 	std::ostringstream out;
 };
 
 
 
+// Raw write for numeric types.
+template <class C>
+DataWriter &DataWriter::WriteRaw(const C &c)
+{
+	static_assert(std::is_arithmetic<C>::value,
+		"DataWriter cannot output anything but strings and arithmetic types.");
+	out << c;
+	return *this;
+}
+
+
+
 // The Write() function can take any number of arguments, each of which becomes
 // a token. They must be either strings or numeric types.
 template <class A, class ...B>
-void DataWriter::Write(const A &a, B... others)
+DataWriter &DataWriter::Write(const A &a, B... others)
 {
 	WriteToken(a);
 	Write(others...);
+	return *this;
 }
 
 
 
 // Write any numeric type as a single token.
 template <class A>
-void DataWriter::WriteToken(const A &a)
+DataWriter &DataWriter::WriteToken(const A &a)
 {
 	static_assert(std::is_arithmetic<A>::value,
 		"DataWriter cannot output anything but strings and arithmetic types.");
 
-	out << *before << a;
-	before = &space;
+	WriteSeparator();
+	out << a;
+	before = &separator;
+	return *this;
+}
+
+// Writes a series of tokens without terminating the line.
+template <class A, class ...B>
+DataWriter &DataWriter::WriteToken(const A &a, B... others)
```

Comment:
This function looks exactly the same as `Write`? or am I missing something

Reply:
Write(), in general, terminates the line.
Write(multiple things) only separates it after all of them are written, not between each one.
WriteToken() doesn't add line separators.

This is relatively consistent with the previous behaviour.

I should edit the comment to clarify that.

Reply:
(Also, Write() seems to be able to handle DataNodes, while WriteToken doesn't. That might mess with the parameter expansion though, not sure how that works right now.)

Reply:
Write also calls WriteTokens though

Reply:
> (Also, Write() seems to be able to handle DataNodes, while WriteToken doesn't. That might mess with the parameter expansion though, not sure how that works right now.)

Okay, here's why it's this way:

Data nodes should **_always_** be followed by a newline. Otherwise, we would risk writing tokens to the last line of the node, which would cause some crazy issues. This is why DataNode is only supported in `Write()`. This also explains why it's only supported as a single argument: if the node is not the last token we write, we would either violate the invariant of not putting newlines between tokens, or we would mess with a data node.

We allow a node to be written _after_ a token, though. Not sure if this is actually used anywhere, but it isn't inherently incorrect, and I could see it being used to e.g. write `not` nodes without fully constructing them. Also, fixing that would require adding yet another differently named write function.

> Write also calls WriteTokens though

It writes a node by writing its top-level elements, then recursing into its children.